### PR TITLE
feat: Advanced parser utility (pydantic)

### PR DESCRIPTION
**Pull request by risenberg-cyberark**
**Issue, if available:** #147, #95

## Description of changes:

Added a new validation module. It has the validator decorator code with 3 envelopes (eventbridge, dynamoDB and custom user) and tests. Also has the eventbridge & dynamoDB schemas. 

**Checklist**

 

* [ ] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

## Breaking change checklist

**RFC issue #95  :

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by risenberg-cyberark** [comment]
open issues:
1. update documentation.
2. add SQS/SNS support
3. decided whether the UX is good enough, i.e the lambda handler gets a new dict with two keys: custom and orig, where orig is the original event unparsed (should it be parsed?) and custom is the pydantic parsed object that user defined.
Note that for the dynamoDB you get a list of dicts where each dict has a new and old objects (new & old stream for each record because dynamodb sends a list of records).
4. how will the user let the decorator know in sns/sqs that the message fields is string or a string that represents a json with a custom schema.
5. add more tests?

File: aws_lambda_powertools/validation/schemas.py
Comment (alexcasalboni) [inline aws_lambda_powertools/validation/schemas.py:8]: Maybe let's use `DynamoDBSchema`?

File: aws_lambda_powertools/validation/validator.py
Comment (alexcasalboni) [inline aws_lambda_powertools/validation/validator.py:54]: Maybe let's use `DynamoDBEnvelope`?

Comment (risenberg-cyberark) [reply aws_lambda_powertools/validation/validator.py:54]: sure, np

File: aws_lambda_powertools/validation/schemas.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/validation/schemas.py:8]: sure, np

Comment (risenberg-cyberark) [reply aws_lambda_powertools/validation/schemas.py:8]: fixed.

File: aws_lambda_powertools/validation/validator.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/validation/validator.py]: fixed.

**Comment by heitorlessa** [comment]
**Notes to self to review on Friday**

* [ ] Customers shouldn't need to import `pydantic` to define a model/schema
* [ ] Revisit Schema, Envelope and Model naming convention - There could be confusion here
* [ ] Customers shouldn't need to import `ValidationError` from `pydantic`
* [ ] Revisit DynamoDB Schema names - Schema, Scheme, Record, and DynamoDBStream might be a better name
* [ ] Revisit DynamoDB streams complexity baseline (failing)
* [ ] Review high level function that parses and validates (not a decorator)
* [ ] Revisit `validator` to return parsed event by default instead of `custom/orig` event keys
* [ ] Revisit `validator` option to not return parsed event but original one (prevent breaking those relying on event being a dictionary)
* [ ] Read docs and derive ideas to simplify usage

**Comment by jplock** [comment]
Is there a possibility of implementing event schema validation using the existing fastjsonschema dependency instead of introducing pydantic?

**Comment by risenberg-cyberark** [comment]
@jplock  yes you can, but pydantic is much more than just a schema validator. It gives the users advanced validation capabilities, more type checking (types that JSON doesnt provide), custom logic checks on values (other than empty or null), validate custom relationships between parameters and more. In this PR , we dont only validate but give back the parsed object (a pydantic dataclass basically).

**Comment by heitorlessa** [comment]
> cc @cakepietoast @alexcasalboni your thoughts are mostly welcome here too

@jplock suppose we were to use Pydantic (parser+validation), how would you see yourself using this feature? If not, what would be helpful?

I'm still awe with Pydantic flexibility, the ability to provide runtime safety for customers, and providing auto-complete for customers on common event sources they typically hop onto multiple doc pages to find the right format. 

**However**,  the UX is not easy to use yet. I'm looking at the PR more carefully tomorrow, and will block Monday afternoon to think about ways to simplify this, as Ran did a great job in pulling this together already.

---
**Pros**

* Parses popular event sources and provides auto-complete to event data attributes instead of a dict
* Allows customers to define validation more easily, including runtime type checker with helpful errors
* Plays nicely with popular IDEs as well as deriving JSON Schemas from models

**Cons**

* Envelop, Schema, and Model are intertwined and a bit confusing to use (UX)
* Requires customers to understand Pydantic to use it effectively (learning curve, which is what we aim to lower with this lib)
* Additional dependency (8.2MB), and need to profile/benchmark cold start impact
    - Perhaps we can make an optional dependency using Extras?
---

I fear we might add a powerful feature that not many customers will be able to leverage it to its full extent. If we can't think of a way to find a middle ground by early next week, this will be a good opportunity to break this into two simpler utilities (parser, validator), and add simplicity to this project Tenets.

**Comment by jplock** [comment]
@heitorlessa i like the idea of event schema validation but pydantic feels too heavy for me. All I’d want to do is validate a schema and get a dict. I wouldn’t personally use the full models. That’s why I thought we could build this feature using fastjsonschema as a first release.

**Comment by michaelbrewer** [comment]
@heitorlessa  - i don't like any having any kind of extra dependencies. Currently with no extra dependencies we are already at 8.3M 

I see a few options:

1. Docs: A directory listing of companion libraries like a `Lambda Powertools Awesome Collection`, with a very wrapper implementation. Like getting Chalice + Powertools to work or  Lambda Proxy + Powertools etc... This could be a great resource of cookbook examples.
2. Extras: An extension utility that works when you opt in for that library dependency, but does not require common use cases to include any extra dependencies. 
3. CLI tool: Powertools CLI tool that generates models from a schema file. Then the CLI tool can be quite bloated and generated rich but lightweight model objects from a schema file with a collection of common usage cases like a API Gateway Request / Response. In fact these generate model objects should not need any extra dependencies, and should be treated as read only resource built as part of the makefile.

I prefer a combination of 1 and 3.

----

Current dependencies:

```text
name                                         summary
-------------------------------------------  --------------------------------------------------------------------------------------------------------------------------------------------------------
aws_lambda_powertools                        Python utilities for AWS Lambda functions including but not limited to tracing, logging and custom metric
├── aws-xray-sdk<3.0.0,>=2.5.0               The AWS X-Ray SDK for Python (the SDK) enables Python developers to record and emit information from within their applications to the AWS X-Ray service.
│   ├── botocore>=1.11.3                     Low-level, data-driven core of boto 3.
│   │   ├── docutils<0.16,>=0.10             Docutils -- Python Documentation Utilities
│   │   ├── jmespath<1.0.0,>=0.7.1           JSON Matching Expressions
│   │   ├── python-dateutil<3.0.0,>=2.1      Extensions to the standard Python datetime module
│   │   │   └── six>=1.5                     Python 2 and 3 compatibility utilities
│   │   └── urllib3<1.26,>=1.20              HTTP library with thread-safe connection pooling, file post, and more.
│   ├── future                               Clean single-source support for Python 3 and 2
│   ├── jsonpickle                           Python library for serializing any arbitrary object graph into JSON
│   │   └── importlib-metadata               Read metadata from Python packages
│   │       └── zipp>=0.5                    Backport of pathlib-compatible object wrapper for zip files
│   └── wrapt                                Module for decorators, wrappers and monkey patching.
├── boto3<2.0,>=1.12                         The AWS SDK for Python
│   ├── botocore<1.18.0,>=1.17.50            Low-level, data-driven core of boto 3.
│   │   ├── docutils<0.16,>=0.10             Docutils -- Python Documentation Utilities
│   │   ├── jmespath<1.0.0,>=0.7.1           JSON Matching Expressions
│   │   ├── python-dateutil<3.0.0,>=2.1      Extensions to the standard Python datetime module
│   │   │   └── six>=1.5                     Python 2 and 3 compatibility utilities
│   │   └── urllib3<1.26,>=1.20              HTTP library with thread-safe connection pooling, file post, and more.
│   ├── jmespath<1.0.0,>=0.7.1               JSON Matching Expressions
│   └── s3transfer<0.4.0,>=0.3.0             An Amazon S3 Transfer Manager
│       └── botocore<2.0a.0,>=1.12.36        Low-level, data-driven core of boto 3.
│           ├── docutils<0.16,>=0.10         Docutils -- Python Documentation Utilities
│           ├── jmespath<1.0.0,>=0.7.1       JSON Matching Expressions
│           ├── python-dateutil<3.0.0,>=2.1  Extensions to the standard Python datetime module
│           │   └── six>=1.5                 Python 2 and 3 compatibility utilities
│           └── urllib3<1.26,>=1.20          HTTP library with thread-safe connection pooling, file post, and more.
└── fastjsonschema<2.15.0,>=2.14.4           Fastest Python implementation of JSON schema
```

**Comment by heitorlessa** [comment]
Thanks @jplock and @michaelbrewer for your inputs, much appreciated. 

We'll review this today with the core team and circle back with a solution - As of now, this doesn't meet our Keep It Lean tenet however useful this is.

**Comment by risenberg-cyberark** [comment]
I'll add my two cents here if you dont mind :)
 I've introduced pydantic to my organisation half a year ago. It's been well received and we have an excellent experience with it. 
It's very easy to define a schema. You dont have to use the bells & whistles. You just inherit from BaseModel class and define parameters with the Python.Typing module. It produces a much readable code than a dict schema. You can also extend it and import and have a repo just for schemas. What I like here is that as an end user I have an option. I can define a basic schema or a more sophisticated schema.

Regarding the usage -I added per @heitorlessa 's request the validate function. It's not the decorator. it's a simpler version that take a schema and returns a parsed object, just like you wanted.

Regarding the extra size, as an end user, if i'm getting good performance and more usability options, i really dont mind.

I think this repo can set a high standard for validation. It's a great opportunity and I believe that pydantic can help that.

**Comment by cakepietoast** [comment]
I'll weigh in on this one too. Firstly, pydantic looks like an awesome library, and I can definitely see it being useful in the context of Lambda. With that said, I feel like this PR in its current state is breaking the "keep it lean" tenet - if I've understood @michaelbrewer correctly above, the pydantic dependency roughly doubles the size of the powertools package. I think it would also mean dedicating a significant part of the powertools documentation to explaining pydantic specific concepts.

We don't spell it out explicitly in the tenets (perhaps we should?), but for me "easing the adoption of best practices" means thinking about how to do that for 80%+ of the users. My feeling is that a json validation based on the already included fastjsonschema would be the simplest implementation here, and have the smallest learning curve for the majority of users. Users who need a more involved validation solution could make use of the pydantic library in their own custom middlewares. To help those users, I like the idea of having companion libraries and some examples documented in a cookbook.

**Comment by jplock** [comment]
@michaelbrewer is it necessary to bundle boto3 in favor of people bundling themselves in their applications?

**Comment by heitorlessa** [comment]
Hi Ran - As we've discussed, we've decided to put this PR on hold while we hear from more customers on their use of Pydantic for parsing and business rules validation. 

I've created a [dedicated feature request](https://github.com/awslabs/aws-lambda-powertools-python/issues/147) for this, and added labels to remind us to revisit this in the next 3 months. 

As of now, we'll work on a simple validation tool that uses JSON Schemas to meet the 80% of customers who are used to this mechanism. We hope to hear from more customers interested in using Pydantic, and recommend customers using the future simple validation tool to take a look at Pydantic in the hope to build awareness and educate on what's possible.

As the usage of Pydantic grows for Lambda functions and we hear from more customers the need for this feature, we'd like to resume collaborating on this PR with you.

We appreciate all efforts you've done so far on pulling this together, and we look forward to chatting soon about this topic in the future.

Thank you

**Comment by heitorlessa** [comment]
Sharing it here the UX we discussed today - We will include the Parser utility in 1.7.0 as a featured new utility ;)

This should:

1. Make the implementation and maintenance easier (tests, etc.)
2. Allow advanced customers to keep their workflows while having access to popular event sources schemas/models
3. Make UX simpler to remember and use

----

### Parsing with built-in envelopes

We'll parse the wrapped event for popular event sources using the data schema provided:

```python
class MyBusiness(BaseModel):
    message: str
    message_id: int

@parser(schema=MyBusiness, envelope=envelopes.EVENTBRIDGE)
def handler(event: MyBusiness, context: LambdaContext):
    # MyBusiness data with no EventBridge data
    # We’ll parse event[“detail”] with the schema provided
    event.message
    event.message_id
```

### Extending schemas

For advanced customers who know Pydantic well, or want to get access to the whole payload plus their data parsed, they will be able to extend official schemas:

```python
class MyBusiness(BaseModel):
    message: str
    message_id: int

class MyBusinessEventBridge(EventBridgeSchema):
    detail: MyBusiness # override detail key with MyBusiness data as opposed to a simple Dict

@parser(schema=MyBusinessEventBridge)
def handler(event: MyBusinessEventBride, context: LambdaContext):
     # Event Bridge fields
     event.account_id
     event.region
     # MyBusiness data inside detail
     event.detail.message
```

**Comment by risenberg-cyberark** [comment]
Does anybody know why I'm getting this build error:
All local linter/format section work just fine on my dev env.

Run make lint
poetry run isort -rc aws_lambda_powertools tests example
Creating virtualenv aws-lambda-powertools-s3ycVuVL-py3.6 in /home/runner/.cache/pypoetry/virtualenvs

[FileNotFoundError]
[Errno 2] No such file or directory
Makefile:13: recipe for target 'format' failed
make: *** [format] Error 1
Error: Process completed with exit code 2.

**Comment by risenberg-cyberark** [comment]
@heitorlessa @cakepietoast @jplock  hey guys, do you have any idea why the makefile doesnt pass the format stage? i'm stuck :(

**Comment by heitorlessa** [comment]
> @heitorlessa @cakepietoast @jplock hey guys, do you have any idea why the makefile doesnt pass the format stage? i'm stuck :(

looking into it..

**Comment by heitorlessa** [comment]
The format is now solved upstream in develop - rebasing should work.

For the Pytest on examples, it's best to completely remove the example folder from the repo, since we now have a cookiecutter for that purpose.

I'll push that change to develop later and ping here for rebasing

**Comment by risenberg-cyberark** [comment]
if anyone knows to reduce the complexity of:
ERROR:xenon:module ‘aws_lambda_powertools/utilities/advanced_parser/envelopes/dynamodb.py’ has a rank of B

I'm open to suggestions.. it's just a pydantic data class.

**Comment by risenberg-cyberark** [comment]
@heitorlessa  all ready for ya ;)

File: aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py:28]: docs: could you add a comment as a context for this validation?

File: aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py:35]: docs: could you add a comment as a context for this validation?

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py:25]: docs: could you add a comment as a context for why we need this? one could also pass `schema=bool()` and would pass and fail down there, so I guess it's not a type check only

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py:30]: docs: could you add a comment as a context for parse_raw for non-pydantic maintainers?

File: tests/functional/parser/test_sqs.py
Comment (heitorlessa) [inline tests/functional/parser/test_sqs.py:11]: docs: could you add a comment as a context as to why someone would pass a schema as a str? should we expect something else?

File: pyproject.toml
Comment (heitorlessa) [inline pyproject.toml:27]: docs: could you add a comment as a context for this additional optional lib?

File: aws_lambda_powertools/utilities/advanced_parser/parser.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/parser.py:36]: TODO

File: aws_lambda_powertools/utilities/advanced_parser/parser.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/parser.py:56]: `return handler(parsed_event, context)`

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/dynamodb.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/envelopes/dynamodb.py:12]: docs: comment or docstring on the return if possible

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/sqs.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/envelopes/sqs.py:12]: docs: comment or docstring on the return if possible

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/event_bridge.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/advanced_parser/envelopes/event_bridge.py:12]: docs: comment or docstring on the return if possible

**Review by heitorlessa — CHANGES_REQUESTED** [review]
New user experience looks GREAT! Added some comments to help us maintain as not all maintainers are well versed with Pydantic, so any additional context helps.

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/envelopes/base.py:25]: added

File: aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py:35]: added

File: aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/schemas/sqs.py:28]: added

File: aws_lambda_powertools/utilities/advanced_parser/schemas/dynamodb.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/schemas/dynamodb.py:17]: added

File: Makefile
Comment (risenberg-cyberark) [reply Makefile:8]: fixed

File: aws_lambda_powertools/utilities/advanced_parser/parser.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/parser.py:56]: fixed

File: aws_lambda_powertools/utilities/advanced_parser/parser.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/parser.py:36]: fixed

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/dynamodb.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/envelopes/dynamodb.py:12]: fixed.

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/sqs.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/envelopes/sqs.py:12]: the type hint is good here but i'll add a comment.

File: aws_lambda_powertools/utilities/advanced_parser/envelopes/event_bridge.py
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/advanced_parser/envelopes/event_bridge.py:12]: fixed.

**Comment by heitorlessa** [comment]
**THANK YOU** @risenberg-cyberark for all those changes, UX improvements, and additional comments you've made. 

I'm gonna merge this as-is now. As agreed, I'll make a PR to write the docs for this feature and get your review before merging it. 

As you also spent a ton of effort already, I'll create another PR the following minor changes:

* [ ] **Remove validators**. We need an universal schema owned by Lambda for each event source, or else we'll likely break some customers if a few services change, as we won't be notified.
* [ ] **Snake_case**. IIRC, EventBridge `detailtype` was the only odd one out but I'll double check if there are others
* [ ] **Refactor test event data**. Data_classes and validator utility are partially using the same events, so we'll refactor tests to provide events to all tests as fixtures.
* [ ] **Kwargs over args**. Some internal code is using args over kwargs - Moving to kwargs make it easier to refactor, and order of args wouldn't be impacted.
* [ ] **Docstrings**. Enhance Example section in docstrings.
* [ ] **Remove flake8 polyfill**. Double check whether a fresh Python env hits the issue you hit with that, since we don't use it
* [ ] **Raise own Validation Exception**. For customers wanting to catch that exception (e.g. custom middlewares) they'll have to find Pydantic exception to catch it properly
* [ ] **Enhance imports**. Envelopes and a few others can be turned into high level imports to clean up customers' code, plus DX.
* [ ] **Test coverage**. Validation errors aren't being tested as of now.
* [ ] **[Optional] Enhancing envelope selection**. Time allowing, I'll experiment with providing a similar experience we have in the `validator` envelope utility `envelope.<event_source>`.
* [ ] **[Optional] Create a standalone parser**. Time allowing, I'll experiment creating a standalone `parser` function with the same functionality the decorator
* [ ] **[Optional] Support for API Gateway, Kinesis, S3, and SNS**. Likely to be after 1.7.0, but time allowing we can do that sooner.

**Review by heitorlessa — APPROVED** [review]

**Comment by risenberg-cyberark** [comment]
@heitorlessa can you maybe put them in comments? i'd hate for that logic to be forgotten. These are connections that json schemas cant really define (only documentation).

**Comment by heitorlessa** [comment]
> @heitorlessa can you maybe put them in comments? i'd hate for that logic to be forgotten. These are connections that json schemas cant really define (only documentation).

Sure thing - Commented them out, and added a link to this discussion for history purposes.

**Comment by mwarkentin** [comment]
@heitorlessa are the extras required for this functionality included in the SAR app / lambda layer? Or do those need to be handled separately? Might be worth adding a note in the docs, not sure if that would be under the parser docs or the lambda layer installation docs?

**Comment by heitorlessa** [comment]
> @heitorlessa are the extras required for this functionality included in the SAR app / lambda layer? Or do those need to be handled separately? Might be worth adding a note in the docs, not sure if that would be under the parser docs or the lambda layer installation docs?

Hey @mwarkentin - the latter, handled separately as described in the parser docs.

Unsure whether it worth creating a separate Layer for that tbh given the additional operational overhead. In V2, we'd like to create each utility to be pip installable if a theory works, then we could consider multiple layers I suppose.

Happy to discuss more in a separate issue tho ;)

**Pull request merged**
